config/pico/config.cpp_doc.md
Based on the summary of the code, I can infer that the following files may contain relevant information:

1. comms/B0XXInputViewer.hpp: This file likely contains the definition of the B0XXInputViewer class, which is used as a communication backend in the code.
2. comms/DInputBackend.hpp: This file likely contains the definition of the DInputBackend class, which is another communication backend.
3. comms/GamecubeBackend.hpp: This file likely contains the definition of the GamecubeBackend class, which is yet another communication backend.
4. comms/N64Backend.hpp: This file likely contains the definition of the N64Backend class, which is another communication backend.
5. comms/NintendoSwitchBackend.hpp: This file likely contains the definition of the NintendoSwitchBackend class, which is another communication backend.
6. comms/XInputBackend.hpp: This file likely contains the definition of the XInputBackend class, which is another communication backend.
7. config/mode_selection.hpp: This file likely contains the definitions and settings for different input modes.
8. core/CommunicationBackend.hpp: This file likely contains the definition of the CommunicationBackend class, which is the base class for all communication backends.
9. core/InputMode.hpp: This file likely contains the definition of the InputMode class, which represents different input modes.
10. core/KeyboardMode.hpp: This file likely contains the definition of the KeyboardMode class, which represents the keyboard input mode.
11. core/pinout.hpp: This file likely contains the pin mappings used in the code.
12. core/socd.hpp: This file likely contains the definition of the socd namespace, which may have utility functions for handling simultaneous input directions.
13. core/state.hpp: This file likely contains the definition of the InputState struct, which represents the current state of the input buttons.
14. input/GpioButtonInput.hpp: This file likely contains the definition of the GpioButtonInput class, which handles reading button inputs from GPIO pins.
15. input/NunchukInput.hpp: This file likely contains the definition of the NunchukInput class, which handles reading input from a Nunchuk controller.
16. joybus_utils.hpp: This file likely contains utility functions related to Joybus communication.
17. modes/Melee20Button.hpp: This file likely contains the definition of the Melee20Button class, which represents the input mode for playing Super Smash Bros. Melee with 20XX mods.
18. modes/ProjectM.hpp: This file likely contains the definition of the ProjectM class, which represents the input mode for playing Project M.
19. modes/Smash64.hpp: This file likely contains the definition of the Smash64 class, which represents the input mode for playing Super Smash Bros. for Nintendo 64.
20. modes/WingmanFgcMode.hpp: This file likely contains the definition of the WingmanFgcMode class, which represents a specific input mode for the Brook Wingman controller.
21. modes/Ultimate.hpp: This file likely contains the definition of the Ultimate class, which represents the input mode for playing Super Smash Bros. Ultimate.
22. modes/FgcMode.hpp: This file likely contains the definition of the FgcMode class, which represents a generic Fighting Game Community (FGC) input mode.
23. stdlib.hpp: This file likely contains standard library includes and definitions used in the code.

Please note that these are just educated guesses based on the provided code snippet, and the actual file names may differ in your codebase.

HAL/pico/include/tusb_config_pico.h_doc.md
Based on the summary, this file seems to contain the configuration settings for the RP2040 microcontroller in a USB project. It defines the mode (device or host), enables different USB stacks (device and host), sets the endpoint sizes for various USB classes (CDC, MSC, HID, MIDI, Vendor), and specifies the buffer sizes for data transmission and reception in each class. Additionally, it includes settings for the host configuration, such as the buffer size for enumeration and the maximum number of devices supported.

config/pico/config.cpp
#include "comms/B0XXInputViewer.hpp"
#include "comms/DInputBackend.hpp"
#include "comms/GamecubeBackend.hpp"
#include "comms/N64Backend.hpp"
#include "comms/NintendoSwitchBackend.hpp"
#include "comms/XInputBackend.hpp"
#include "config/mode_selection.hpp"
#include "core/CommunicationBackend.hpp"
#include "core/InputMode.hpp"
#include "core/KeyboardMode.hpp"
#include "core/pinout.hpp"
#include "core/socd.hpp"
#include "core/state.hpp"
#include "input/GpioButtonInput.hpp"
#include "input/NunchukInput.hpp"
#include "joybus_utils.hpp"
#include "modes/Melee20Button.hpp"
#include "modes/ProjectM.hpp"
#include "modes/Smash64.hpp"
#include "modes/WingmanFgcMode.hpp"
#include "modes/Ultimate.hpp"
#include "modes/FgcMode.hpp"
#include "stdlib.hpp"


#include <pico/bootrom.h>

CommunicationBackend **backends = nullptr;
size_t backend_count;
KeyboardMode *current_kb_mode = nullptr;

GpioButtonMapping button_mappings[] = {
    {&InputState::l,            5 },
    { &InputState::left,        4 },
    { &InputState::down,        3 },
    { &InputState::right,       2 },
    { &InputState::w,         1 },
    { &InputState::mod_x,       6 },
    { &InputState::mod_y,       7 },

    { &InputState::select,      10},
    { &InputState::start,       0 },
    { &InputState::home,        11},
    { &InputState::nunchuk_c,   8 }, // Dpad Toggle button
    { &InputState::c_left,      13},
    { &InputState::c_up,        12},
    { &InputState::c_down,      15},
    { &InputState::a,           14},
    { &InputState::c_right,     16},
    { &InputState::b,           26},
    { &InputState::x,           21},
    { &InputState::z,           19},
    { &InputState::up,          17},

    { &InputState::r,           27},
    { &InputState::y,           22},
    { &InputState::lightshield, 20},
    { &InputState::midshield,   18},
};
size_t button_count = sizeof(button_mappings) / sizeof(GpioButtonMapping);

const Pinout pinout = {
    .joybus_data = 28,
    .mux = -1,
    .nunchuk_detect = -1,
    .nunchuk_sda = -1,
    .nunchuk_scl = -1,
};

void setup() {
    // Create GPIO input source and use it to read button states for checking button holds.
    GpioButtonInput *gpio_input = new GpioButtonInput(button_mappings, button_count);

    InputState button_holds;
    gpio_input->UpdateInputs(button_holds);

    // Bootsel button hold as early as possible for safety.
    if (button_holds.start) {
        reset_usb_boot(0, 0);
    }

    // Turn on LED to indicate firmware booted.
    gpio_init(PICO_DEFAULT_LED_PIN);
    gpio_set_dir(PICO_DEFAULT_LED_PIN, GPIO_OUT);
    gpio_put(PICO_DEFAULT_LED_PIN, 1);

    // Create array of input sources to be used.
    static InputSource *input_sources[] = { gpio_input };
    size_t input_source_count = sizeof(input_sources) / sizeof(InputSource *);

    ConnectedConsole console = detect_console(pinout.joybus_data);

    /* Select communication backend. */
    CommunicationBackend *primary_backend;
    if (console == ConnectedConsole::NONE) {
        if (button_holds.x) {
            // If no console detected and X is held on plugin then use Switch USB backend.
            NintendoSwitchBackend::RegisterDescriptor();
            backend_count = 1;
            primary_backend = new NintendoSwitchBackend(input_sources, input_source_count);
            backends = new CommunicationBackend *[backend_count] { primary_backend };
            // Default to Wingman FGC mode upon plugin to Brook Wingman.
            primary_backend->SetGameMode(new WingmanFgcMode(socd::SOCD_NEUTRAL, socd::SOCD_NEUTRAL));
            return;
        } else if (button_holds.z) {
            // If no console detected and Z is held on plugin then use DInput backend.
            TUGamepad::registerDescriptor();
            TUKeyboard::registerDescriptor();
            backend_count = 2;
            primary_backend = new DInputBackend(input_sources, input_source_count);
            backends = new CommunicationBackend *[backend_count] {
                primary_backend, new B0XXInputViewer(input_sources, input_source_count)
            };
        } else {
            // Default to XInput mode if no console detected and no other mode forced.
            backend_count = 2;
            primary_backend = new XInputBackend(input_sources, input_source_count);
            backends = new CommunicationBackend *[backend_count] {
                primary_backend, new B0XXInputViewer(input_sources, input_source_count)
            };
            primary_backend->SetGameMode(
                new Melee20Button(socd::SOCD_2IP_NO_REAC, { .crouch_walk_os = false })
                //new FgcMode(socd::SOCD_NEUTRAL, socd::SOCD_NEUTRAL)
            );
        }
    } else {
        if (console == ConnectedConsole::GAMECUBE) {
            primary_backend =
                new GamecubeBackend(input_sources, input_source_count, pinout.joybus_data);
                primary_backend->SetGameMode(
                    new Melee20Button(socd::SOCD_2IP_NO_REAC, { .crouch_walk_os = false })
                    //new Ultimate(socd::SOCD_2IP)
                );
        } else if (console == ConnectedConsole::N64) {
            primary_backend = new N64Backend(input_sources, input_source_count, pinout.joybus_data);
            primary_backend->SetGameMode(new Smash64(socd::SOCD_NEUTRAL, socd::SOCD_NEUTRAL));

        }

        // If console then only using 1 backend (no input viewer).
        backend_count = 1;
        backends = new CommunicationBackend *[backend_count] { primary_backend };
    }

    // Default to Melee mode.
    
}

void loop() {
    select_mode(backends[0]);

    for (size_t i = 0; i < backend_count; i++) {
        backends[i]->SendReport();
    }

    if (current_kb_mode != nullptr) {
        current_kb_mode->SendReport(backends[0]->GetInputs());
    }
}

/* Nunchuk code runs on the second core */
NunchukInput *nunchuk = nullptr;

void setup1() {
    while (backends == nullptr) {
        tight_loop_contents();
    }

    // Create Nunchuk input source.
    nunchuk = new NunchukInput(Wire, pinout.nunchuk_detect, pinout.nunchuk_sda, pinout.nunchuk_scl);
}

void loop1() {
    if (backends != nullptr) {
        nunchuk->UpdateInputs(backends[0]->GetInputs());
        busy_wait_us(50);
    }
}
